### PR TITLE
added failing test for ignore_older with file rename

### PR DESCRIPTION
**Pull request by pghalliday**
Just adding this for illustrative purposes related to #601 - not expecting a failing test to be merged.
Once the cause is identified I suspect a better test could be formulated

### CODE REVIEW

File: filebeat/tests/system/test_prospector.py [hunk 1]
```diff
@@ -103,3 +103,56 @@ def test_rotating_ignore_older_larger_write_rate(self):
...
+            file.write("Line {}\n".format(lines))
+
+        # wait for log to be read
+        self.wait_until(
+            lambda: self.output_has(lines=lines),
+            max_timeout=15)
+
+        # log rotate
+        os.rename(testfile, testfile + ".1")
+        open(testfile, 'w').close()
```
Comment (pghalliday) [inline filebeat/tests/system/test_prospector.py]: This empty file creation seems to not be necessary but maybe should be

File: filebeat/tests/system/test_prospector.py [hunk 2]
```diff
@@ -103,3 +103,56 @@ def test_rotating_ignore_older_larger_write_rate(self):
...
+
+        # wait for file to be closed due to ignore_older
+        self.wait_until(
+            lambda: self.log_contains(
+              "Error: Stop harvesting as file is older then ignore_older: {};".format(os.path.abspath(testfile))),
+            max_timeout=10)
+        time.sleep(2)
```
Comment (pghalliday) [inline filebeat/tests/system/test_prospector.py]: This timeout is actually what causes the test to fail - I think because it allows the second ignore older message for the same file to come through, after which things get into a non working state

**Comment by pghalliday** [comment]
Note that whether the empty file is created or not I still get 2 ignore_older messages for `test.log` and none for `test.log.1`

```
12:44 $ cat build/system-tests/run/test_prospector.Test.test_rotating_ignore_older_low_write_rate/filebeat.log | grep ERR
2015/12/31 11:43:06.641049 log.go:102: ERR File reading error. Stopping harvester. Error: Stop harvesting as file is older then ignore_older: /Users/pghalliday/projects/go/src/github.com/elastic/beats/filebeat/build/system-tests/run/test_prospector.Test.test_rotating_ignore_older_low_write_rate/log/test.log; Last change was: 1.024314312s
2015/12/31 11:43:06.955414 log.go:102: ERR File reading error. Stopping harvester. Error: Stop harvesting as file is older then ignore_older: /Users/pghalliday/projects/go/src/github.com/elastic/beats/filebeat/build/system-tests/run/test_prospector.Test.test_rotating_ignore_older_low_write_rate/log/test.log; Last change was: 1.036009208s
```

File: filebeat/tests/system/test_prospector.py (hunk 1 above)
Comment (ruflin) [inline filebeat/tests/system/test_prospector.py]: It seems like the issue is here somewhere. If I replace this by the code below, the test passed:

```
        with open(testfile, 'a') as file:
            file.write("Line {}\n".format(lines))
            file.close()
```

It seems like something does not work as intended if an empty file is read.

Comment (ruflin) [inline filebeat/tests/system/test_prospector.py]: It definitively seems to be related to the empty file. I added one more test without file rotation and just starting with an empty file. The same issue happens.

Comment (ruflin) [inline filebeat/tests/system/test_prospector.py]: Ignore the last comment about without rotation. Seems like that was some caching issue on my side. Looks good now :-(

**Comment by ruflin** [comment]
Now it seems again more like a race condition. After running the tests, the registrar file has the following content:

```
{"/Users/ruflin/Dev/gopath/src/github.com/elastic/beats/filebeat/build/system-tests/run/test_prospector.Test.test_rotating_ignore_older_low_write_rate/log/test.log":{"source":"/Users/ruflin/Dev/gopath/src/github.com/elastic/beats/filebeat/build/system-tests/run/test_prospector.Test.test_rotating_ignore_older_low_write_rate/log/test.log","offset":7,"FileStateOS":{"inode":12897236,"device":16777220}}}
```

Checking the inodes of the log files outputs the following:

```
12897239 -rw-r--r--  1 ruflin  staff     7B Dec 31 15:31 test.log
12897236 -rw-r--r--  1 ruflin  staff    25B Dec 31 15:31 test.log.1
```

The registrar file should have two entries. One for the rotate file and one for the new. And the content in the registrar file is somehow messy, as it has the path of the the new file but the inode of the `test.log.1` file. 

One more part that is out of order can be found in the log below. First it finds correctly the rotate file and does `Launching harvester on rotated file`. After that if it finds the new file `Start harvesting unknown file` but then detects again a file renaming in `File rename was detected`. The problem seems to be in the prospector [here](https://github.com/elastic/beats/blob/1.0.1/filebeat/crawler/prospector.go#L309) inside the Prospector.getPreviousFile() function. This should return false, but returns true.

```
prospector.go:216: DBG  Start next scan
prospector.go:241: DBG  scan path /Users/ruflin/Dev/gopath/src/github.com/elastic/beats/filebeat/build/system-tests/run/test_prospector.Test.test_rotating_ignore_older_low_write_rate/log/*
prospector.go:255: DBG  Check file for harvesting: /Users/ruflin/Dev/gopath/src/github.com/elastic/beats/filebeat/build/system-tests/run/test_prospector.Test.test_rotating_ignore_older_low_write_rate/log/test.log
prospector.go:383: DBG  Update existing file for harvesting: /Users/ruflin/Dev/gopath/src/github.com/elastic/beats/filebeat/build/system-tests/run/test_prospector.Test.test_rotating_ignore_older_low_write_rate/log/test.log
prospector.go:408: DBG  Launching harvester on rotated file: /Users/ruflin/Dev/gopath/src/github.com/elastic/beats/filebeat/build/system-tests/run/test_prospector.Test.test_rotating_ignore_older_low_write_rate/log/test.log
prospector.go:255: DBG  Check file for harvesting: /Users/ruflin/Dev/gopath/src/github.com/elastic/beats/filebeat/build/system-tests/run/test_prospector.Test.test_rotating_ignore_older_low_write_rate/log/test.log.1
prospector.go:308: DBG  Start harvesting unknown file: /Users/ruflin/Dev/gopath/src/github.com/elastic/beats/filebeat/build/system-tests/run/test_prospector.Test.test_rotating_ignore_older_low_write_rate/log/test.log.1
prospector.go:348: DBG  File rename was detected: /Users/ruflin/Dev/gopath/src/github.com/elastic/beats/filebeat/build/system-tests/run/test_prospector.Test.test_rotating_ignore_older_low_write_rate/log/test.log -> /Users/ruflin/Dev/gopath/src/github.com/elastic/beats/filebeat/build/system-tests/run/test_prospector.Test.test_rotating_ignore_older_low_write_rate/log/test.log.1
log.go:74: INFO Harvester started for file: /Users/ruflin/Dev/gopath/src/github.com/elastic/beats/filebeat/build/system-tests/run/test_prospector.Test.test_rotating_ignore_older_low_write_rate/log/test.log
log.go:190: DBG  harvest: "/Users/ruflin/Dev/gopath/src/github.com/elastic/beats/filebeat/build/system-tests/run/test_prospector.Test.test_rotating_ignore_older_low_write_rate/log/test.log" (offset snapshot:0)
spooler.go:108: DBG  Flushing spooler because of timemout. Events flushed: 0
```

**Comment by ruflin** [comment]
Interestingly when my computer was under load, the test started to pass. It fails most of the time, but sometimes it seems to be flaky. So my current theory is as following. The problem lies in the check of previous files in [getPreviousFiles()](https://github.com/elastic/beats/blob/1.0.1/filebeat/crawler/prospector.go#L394) an related to this prospectorList and missingFiles. It seems like states here are overwritten by each other when the file is empty. When I file is empty Harvesting takes much less time as no events have to be sent which leads to this race condition.

I think through our reorganisation for 1.1 this problem was solved because we removed some of these race conditions. I would argue it is somehow an edge case, as I would assume a new log file that is generated always directly has content inside (otherwise it would not have to be generated). 

So my suggestion is that people which hit this issue start to use the 1.1.0 nightly build which can be found [here](https://beats-nightlies.s3.amazonaws.com/index.html?prefix=filebeat/). @tsg FYI

@pghalliday Let me know if that works for you?

**Comment by pghalliday** [comment]
Cool i'll certainly check out the nightly :)

As far as edge case goes, I'm not so sure. Generally this is how logrotate works.

1. rename file
2. reload/restart service (so it stop using the renamed file which it likely had open)

This usually results in a log file being opened by the service (I might be generalizing but I'm pretty sure it's more common than just nginx, could check out apache too to see if it does the same thing with access logs) thus creating an empty file even before anything is logged

**Comment by pghalliday** [comment]
Just thinking about it passing under load. I had to wait the additional 2 seconds after the initial ignore_older kicks in before the issue occurred. Under load it's possible you have to wait longer and the issue may still occur - I haven't checked though

**Comment by ruflin** [comment]
@pghalliday You could be right, but I still prefer to point people to 1.1 snapshot as the changes are quite major.

Would you mind open your PR agains master so we can merge it in?

**Comment by pghalliday** [comment]
#601 has a test that's compatible with master - it's missing the timeout though. I'll add the timeout there and then you can merge that (the test will still pass though)

**Comment by ruflin** [comment]
@pghalliday Sounds good to me. It means we will not break it again :-)

**Comment by ruflin** [comment]
Closing because https://github.com/elastic/beats/pull/601 was merged.
